src/main/java/org/corfudb/util/serializer/Serializers.java

```diff
@@ -1,39 +1,49 @@
 package org.corfudb.util.serializer;
 
-import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 
-import java.util.Arrays;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.function.Function;
-import java.util.stream.Collectors;
 
 /**
  * Created by mwei on 1/8/16.
  */
 
 @Slf4j
 public class Serializers {
-    public static final Map<Byte, SerializerType> typeMap =
-            Arrays.stream(SerializerType.values())
-                    .collect(Collectors.toMap(SerializerType::asByte, Function.identity()));
 
+    public static final int SYSTEM_SERIALIZERS_COUNT = 10;
+    public static final ISerializer CORFU = new CorfuSerializer((byte) 0);
+    public static final ISerializer JAVA = new JavaSerializer((byte) 1);
+    public static final ISerializer JSON = new JSONSerializer((byte) 2);
+    public static final ISerializer PRIMITIVE = new PrimitiveSerializer((byte) 3);
+
+    private static final Map<Byte, ISerializer> serializersMap;
+    static
+    {
+        serializersMap = new HashMap();
+        serializersMap.put(CORFU.getType(), CORFU);
+        serializersMap.put(JAVA.getType(), JAVA);
+        serializersMap.put(JSON.getType(), JSON);
+        serializersMap.put(PRIMITIVE.getType(), PRIMITIVE);
+    }
 
-    public static final Map<SerializerType, ISerializer> serializerCache = new HashMap<>();
+    private static final Map<Byte, ISerializer> customSerializers = new HashMap<>();
 
-    public static synchronized void setCustomSerializer (Class<? extends ISerializer> serializer) {
-        SerializerType.CUSTOM.setSerializer(serializer);
+    public static ISerializer getSerializer(Byte type) {
+        if (type <= SYSTEM_SERIALIZERS_COUNT) {
+            return serializersMap.get(type);
+        } else {
+            return customSerializers.get(type);
+        }
     }
 
-    public static ISerializer getSerializer(SerializerType type) {
-        return serializerCache.computeIfAbsent(type, x -> {
-            try {
-                return type.getSerializer().newInstance();
-            } catch (InstantiationException | IllegalAccessException ie) {
-                log.error("Error creating a serializer for {}", type.name());
-                throw new RuntimeException(ie);
-            }
-        });
+    public static synchronized void registerSerializer(ISerializer serializer) {
+        if(serializer.getType() > SYSTEM_SERIALIZERS_COUNT) {
```

Comment:
Hmm, so just every one map and have no restrictions on the ids?

I feel that is more error-prone, registering a serializer without knowing all the default serializers' ids, you might override a default serialize without intending to. I think part of the problem is that we use multiple serializers
all over the place. It would be easier if we restrict our usage to one serializer internally and then the user can
override that, or register multiple serializers to be used with the smr objects.